funnel/models/event.py

```diff
@@ -97,9 +107,35 @@ class Participant(BaseMixin, db.Model):
 
     __table_args__ = (db.UniqueConstraint('proposal_space_id', 'email'),)
 
+    @classmethod
+    def get(cls, space, email):
+        return cls.query.filter_by(proposal_space=space, email=email).one_or_none()
+
+    @classmethod
+    def upsert(cls, **fields):
+        participant = cls.get(fields['proposal_space'], fields['email'])
+        if participant:
+            participant._set_fields(fields)
+        else:
+            participant = cls(**fields)
+            db.session.add(participant)
+        return participant
+
+    def update_events(self, events, cancel=False):
```

Comment:
Split this into two methods. I had to read the code to understand that `cancel` reverses the operation.